### DISCUSSION TITLE
Plan to implement multiple upload with S3 ?

### DISCUSSION THREAD

**Discussion by vwasteels** [livewire/livewire#4569] [1. Help]
Hello,

I am implementing `spatie/laravel-medialibrary` component with pro version that offers a livewire components for media uploads.

I am stuck now because of this exception : `S3DoesntSupportMultipleFileUploads `

I have 2 questions :
- will this be implemented in future livewire releases ?
- is there a work around that would allow me to achieve this ? I have tested to change the `temporary_file_upload.disk` to `local` but it seems that it handles both temporary uploads and final uploads with the same disk and eveything ends up on `local`

Thank you :)

**Comment by joshhanley** [comment]
@vwasteels no plans that I know of it implement it as yet.

Can you just use single file uploads?

Otherwise you could upload to local for temp uploads but save to S3. You will have to specify the disk when you go to save though if you want it to be different. If that wasn't working for you, can you share your code you were using so we can have a look?

**Reply by vwasteels** [reply]
Hello, thanks for the answer and sorry for long delay, 
I can't use single upload, in fact I dont have much control on this since it's Spatie MediaLibrary PRO that handles that part for me.
and since it's a private repo, I can't show you the code, I will ask them to check for your solution,
Thanks again !

**Comment by levipeto** [comment]
I am just here to express, that this is a huge pain. 
Is there a reason why multiple uploads to S3 are not supported? What could make that happen?
Thank you!

**Reply by rickmills** [reply]
Also here to say the same. We were moving a site's asset management from local to S3 and this is a show stopper. Really surprised this hasn't been implemented, or at least hasn't got a documented way of working around it without reinventing the wheel.

**Comment by fikrimastor** [comment]
Hi in case if anyone facing the problem, I upload the file into temporary folder by point the `temporary_file_upload.disk` to `tmp`. After that do for loop and upload it into S3. It works.

Initially I'm facing a problem in vapor, which does not have local folder, but tmp folder.

**Comment by aronquiray** [comment]
I found a solution for this. 

If the 'multiple' is enabled, then just overide the trait 
'startUpload' 
```
/**
     * @param  string  $name
     * @param  array<string>  $fileInfo
     * @param  bool  $isMultiple
     * @return void
     */
    public function startUpload(string $name, array $fileInfo, bool $isMultiple): void
    {
        $this->emit('upload:generatedSignedUrl', $name, (new \Livewire\GenerateSignedUploadUrl)->forLocal())->self();
    }
```
on your livewire config 
```
'temporary_file_upload' => [
        'disk' => env('FILESYSTEM_DISK', 'local'),        // Example: 'local', 's3'              Default: 'default'
        'rules' => null,       // Example: ['file', 'mimes:png,jpg']  Default: ['required', 'file', 'max:12288'] (12MB)
        'directory' => 'tmp',   // Example: 'tmp'                      Default  'livewire-tmp'
        'middleware' => null,  // Example: 'throttle:5,1'             Default: 'throttle:60,1'
        'preview_mimes' => [   // Supported file types for temporary pre-signed file URLs.
            'png', 'gif', 'bmp', 'svg', 'wav', 'mp4',
            'mov', 'avi', 'wmv', 'mp3', 'm4a',
            'jpg', 'jpeg', 'mpga', 'webp', 'wma',
        ],
        'max_upload_time' => 5, // Max duration (in minutes) before an upload gets invalidated.
    ],
```
just use the filesystem_disk as your disk (to know if it's local or s3) 
it will be uploaded on s3 + it'll generate a preview url 
![Screenshot_321](https://user-images.githubusercontent.com/13714092/212060803-39b7a309-2b9b-406e-ad9a-681c7286fd54.png)
![Screenshot_322](https://user-images.githubusercontent.com/13714092/212060819-13361cc1-81b2-4183-8ead-dd7a00ef9abb.png)

Not sure why it worked but it worked :D 
PS : don't forget to add a lifecycle on your tmp if you're handling number of files.

**Reply by Edmonbelchev** [reply]
That works if you want to upload temp files in your local folder.

`'disk' => env('FILESYSTEM_DISK', 'local')`

Otherwise if you want to store temp files in S3 it won't allow you to upload multiple files.

**Comment by dennisprudlo** [comment]
~@joshhanley Is there any reason why this isn't supported? I would assume there was some weird edge case scenario or security issue and that's why it's not implemented as of now.~

~Would want to try @aronquiray solution by simply overriding the method but don't want to shoot myself in the foot there. I mean there must have been some reason why the `S3DoesntSupportMultipleFileUploads` exception is being thrown.~

Nevermind. I just realized that the above solution simply ignores the S3 Upload configuration and uses the local upload URL.

**Reply by aronquiray** [reply]
yep. it uses the 'forLocal' but when you update this line ` 'disk' => env('FILESYSTEM_DISK', 'local'),`, I think it assumes it's local but it's not 😅

**Reply by dennisprudlo** [reply]
The file **ends up** in your bucket, however it's uploaded to your server, then your server moves the file to the bucket. But the whole purpose of the feature is, that the file is uploaded directly to S3 avoiding load on your server.

Here's how: This part generates a *local* upload URL:
```php
(new \Livewire\GenerateSignedUploadUrl)->forLocal())
```
It simply generates a signed URL that expires after some time (The code for the S3 upload URL is right beneath that function).
https://github.com/livewire/livewire/blob/2c4c29dc26b9bf248543ef05c1f820186184baa1/src/GenerateSignedUploadUrl.php#L10-L15

The URL is going to be `https://your-app.com/livewire/upload-file` and points to your server. This is the URL the file (or files) will be uploaded to. The route is registered here:
https://github.com/livewire/livewire/blob/fa0441bf82f1674beecb3a8ad8a4ae428736ed18/src/LivewireServiceProvider.php#L149-L151

The route action is `[FileUploadHandler::class, 'handle']`, so let's have a look. The `handle` method check the signature and get the disk, then validates and uploads the file:
https://github.com/livewire/livewire/blob/0a878717bec8cd94aa3b30bccb905eaeef98dfd2/src/Controllers/FileUploadHandler.php#L30-L46

And the magic happens here:
https://github.com/livewire/livewire/blob/0a878717bec8cd94aa3b30bccb905eaeef98dfd2/src/Controllers/FileUploadHandler.php#L39-L41
The disk is specified as `s3` as you said. So the filesystem moves the file to whatever is configured as the `s3` disk -> which probably is your S3 bucket. So the files are first being uploaded to your server, then your server moves the files to the bucket, but it's not a direct upload.

The URL ultimately defines where the file is being uploaded to
- `https://roublez.test/livewire/upload-file?expires=1675850806&signature=abcdef`
- `https://roublez-staging-space.fra1.digitaloceanspaces.com/livewire-tmp?X-Amz-Expires=300&X-Amz-Signature=abcdef` (and some other query "X-Amz" params)

**Comment by mc0657** [comment]
I was using local driver for tmp upload but there is the [issue with the long filenames](https://github.com/livewire/livewire/issues/2834). That's why I switched to S3 and I had to give up using multiple upload.

The upload management looks like a real weakness of Livewire. I hope the problems will be fixed soon, at least on one of the drivers.

**Comment by woodspire** [comment]
@calebporzio With Livewire v3 released, does it currently support multi-file uploads directly to s3 ? Can it be fixed now with the new code architecture ?

**Comment by stefro** [comment]
I also came up with the same solution as @aronquiray before I stumbled on this discussion. I found this to be working very well: the file is uploaded to S3 in the `livewire-tmp` dir, and I have access to the array of `TemporaryUploadedFile`'s in the Livewire component.

But I can't seem to figure out **WHY** Livewire is trying create these pre-signed urls. Where are they used? I want to proceed with overriding the `startUpload` method, but I have a feeling I'm missing something important. Can anyone explain this to me?

**Reply by dennisprudlo** [reply]
Were those files uploaded directly to S3 with what you did or were they uploaded to the server and then moved to the bucket?

When I tried this the files "ended up" in the bucket but when through my server.

**Reply by stefro** [reply]
As soon as I select the files and click `open` in the chrome dialog box, the files are uploaded to the `livewire-tmp` folder on S3. I can't find any evidence that they're also stored in my `Storage` folder on my server. 

@dennisprudlo, how do you know they went trough your server?

**Reply by dennisprudlo** [reply]
I simply checked the upload url in my browsers inspector. In my case it should be a DigitalOcean Spaces url but the host was my app's domain.

I tried to explain how the upload URL is generated here: https://github.com/livewire/livewire/discussions/4569#discussioncomment-4903376

**Reply by stefro** [reply]
I'm using the `$file->temporaryUrl()` to get the url and this is a url directly on S3. 

I'm confused, why this doesn't work “out of the box”. By overriding the `startUpload` method as described above it uploads the files directly to S3 and with `$file->temporaryUrl()` on each file I can also preview them. Why is Livewire by default throwing an `S3DoesntSupportMultipleFileUploads` exception. What am I missing?

**Comment by lrljoe** [comment]
I mean, you could mount an S3 bucket as a local filesystem for uploads. Not ideal tho.  Pretty sure this is a limitation with the flysystem driver (but I could be wrong).

**Comment by stefro** [comment]
I'm writing this as a follow-up.  As I've mentioned before, I was successful in using the method as described above (https://github.com/livewire/livewire/discussions/4569#discussioncomment-4666890).

After a while we came across errors in production when users were trying to upload bigger files. There was no exact size where the errors started, sometimes at 6 MB, sometimes at 10. Then, I came across this awesome post: https://fly.io/laravel-bytes/multi-file-upload-livewire/.

They are not binding the file input field to the model, but create a separate (multiple file upload) field instead. When this field is changed, it will send the files one by one to the model. This allows us to upload multiple files to S3 with Livewire without overriding the default `startUpload` method.

**Reply by aronquiray** [reply]
I'll try to look into the link you provide @stefro . I moved on to this problem as the files we're uploading are just small in size. If you're using a serverless setup like Vapor. I think it's the chunk size problem as it limits up to 4mb files.

**Reply by stefro** [reply]
@aronquiray The article also describes how chunking would work. I'm not using it in Vapor, but I'm curious to know if you will be running into issues while using this technique. Please keep us informed!

**Comment by phuclh** [comment]
Does Livewire 3 support multiple uploads for tmp files on S3?

**Comment by spizzo14** [comment]
anyway we can get a PR merged for Livewire 3 to allow what @phuclh  said? Willing to pay a premium rate!

**Comment by d-shannon** [comment]
A solution to this problem that has been working for me is to use Filepond. Livewire doesn't detect if Filepond input is accepting multiple files, and the files get uploaded to s3 without issue.

I picked up a working component from here - https://rappasoft.com/blog/creating-a-filepond-component-using-blade-livewire-alpinejs-then-validating-storing-with-spatie-media-library

**Reply by AndrewMonty** [reply]
> A solution to this problem that has been working for me is to use Filepond. Livewire doesn't detect if Filepond input is accepting multiple files, and the files get uploaded to s3 without issue.
> 
> I picked up a working component from here - https://rappasoft.com/blog/creating-a-filepond-component-using-blade-livewire-alpinejs-then-validating-storing-with-spatie-media-library

Thanks for this. Ran into this issue today when trying out Laravel Cloud, can't use the local disk for preview work-around there either. Filepond seems like it avoids the problem by making individual calls to `/livewire/upload-file` for each selected file.

I would prefer a more customizable solution, but the UX of FilePond is pretty good so I'll run with it for now.

Here's my `<x-filepond />` blade component working with Livewire 3. I opted to install FilePond via npm instead of the CDN.

```php
<div wire:ignore x-data x-cloak x-init="
    () => {
        const post = FilePond.create($refs.{{ $attributes->get('ref') ?? 'input' }});
        post.setOptions({
            allowMultiple: {{ $attributes->has('multiple') ? 'true' : 'false' }},
            server: {
                process:(fieldName, file, metadata, load, error, progress, abort, transfer, options) => {
                    $wire.upload('{{ $attributes->whereStartsWith('wire:model')->first() }}', file, load, error, progress)
                },
                revert: (filename, load) => {
                    $wire.removeUpload('{{ $attributes->whereStartsWith('wire:model')->first() }}', filename, load)
                },
            },
            allowImagePreview: {{ $attributes->has('allowFileTypeValidation') ? 'true' : 'false' }},
            imagePreviewMaxHeight: {{ $attributes->has('imagePreviewMaxHeight') ? $attributes->get('imagePreviewMaxHeight') : '256' }},
            allowFileTypeValidation: {{ $attributes->has('allowFileTypeValidation') ? 'true' : 'false' }},
            acceptedFileTypes: {!! $attributes->get('acceptedFileTypes') ?? 'null' !!},
            allowFileSizeValidation: {{ $attributes->has('allowFileSizeValidation') ? 'true' : 'false' }},
            maxFileSize: {!! $attributes->has('maxFileSize') ? "'".$attributes->get('maxFileSize')."'" : 'null' !!}
        });
    }
">
    <input type="file" x-ref="{{ $attributes->get('ref') ?? 'input' }}" />
</div>
```

usage:

```php
<x-filepond
    wire:model="photos"
    multiple
    allowFileTypeValidation
    acceptedFileTypes="['image/png', 'image/jpg', 'image/jpeg']"
    allowFileSizeValidation
    maxFileSize="4mb"
/>
```

**Comment by Edmonbelchev** [comment]
I managed to make it work on a Livewire2 project. I did it in a dirty way, because I couldn't find a proper way to extend the JavaScript that runs the upload to S3. 

Anyway I guess everyone is using Livewire3 that's why I wrote an article on how to implement it on latest version.

https://www.linkedin.com/pulse/implementing-multiple-file-upload-s3-bucket-edmon-belchev-4oagf/?trackingId=tU8tiJuNQkmi2x1dNdNM5g%3D%3D

I've also created a repository with working solution for both Livewire 2 and Livewire 3.

https://github.com/Edmonbelchev/livewire-multiple-file-upload

**Comment by 0hra** [comment]
Bump... This is very relevant if you want to use multiple file upload on Laravel Cloud which does not allow local storage I believe.

**Reply by timgavin** [reply]
Yep, this has turned into a problem for me, especially since I'm getting `client_body_buffer_size` warnings on Laravel Cloud, and there's no way to fix it without uploading directly to S3, which I can't do because of Livewire's current limitations.

**Reply by 0hra** [reply]
@joshhanley, sorry for the tag, but is this on your and Caleb's radar?

**Reply by timgavin** [reply]
Laravel Cloud support helped me set the `client_body_buffer_size`, and I had to bump it to 512k to remove the warnings; that's pretty high. The largest image I uploaded was only 706k. Livewire really needs to allow multiple images to S3.

**Reply by 0hra** [reply]
I've implemented this instead:
https://github.com/spatie/livewire-filepond
With this I can upload directly to a bucket no problem.

**Reply by timgavin** [reply]
Does that handle conversions?

**Comment by spizzo14** [comment]
@0hra i've literally been begging them as well... trying to use any resource i have to get this done. We despreately need this feature

**Comment by marcosgrimm** [comment]
Just saw https://github.com/livewire/livewire/pull/9355, which is excellent!

Are there any dates when it will be available?

@joshhanley

**Reply by spizzo14** [reply]
@joshhanley name your price, we desperately need this

**Reply by lrljoe** [reply]
> @joshhanley name your price, we desperately need this

I believe the current price is $1,750 in a money order that Josh can pick up at a random bank....

In all seriousness - give the PR/Issue a thumbs up to help draw attention to it.,

**Reply by spizzo14** [reply]
@lrljoe whatever it takes, Livewire is critical for our operations

**Reply by lrljoe** [reply]
> @lrljoe whatever it takes, Livewire is critical for our operations

In the interim, using the local temp upload dir, then pushing/storing to S3 works.

Appreciate it's not ideal tho.

The referenced PR having "whitespace fixes" in it does make it harder to review tho, so hopefully the PR author will remove those.

**Comment by wilianmaique** [comment]
No plans to add upload via 'chunks'? Since 99% of sites use Cloudflare but there is a 100mb limit, it would be nice to implement this natively in LiveWire, to send large files in parts.

**Comment by Hesesses** [comment]
Whats the status of this now?

**Reply by wilianmaique** [reply]
> Whats the status of this now?

It's already being implemented; there are several commits.

**Comment by brrrm** [comment]
As @0hra mentioned: https://github.com/spatie/livewire-filepond gives you a working (multi)-file upload component. I just tested it on Laravel Cloud.
